### DISCUSSION TITLE
Plugin marketplace

### DISCUSSION THREAD

**Discussion by charbs-io** [pocketbase/pocketbase#1196] [Q&A] [answered]
This is probably out of scope for 1.0.0, but has there been a consideration to enable a marketplace of sorts that allows users of pocketbase to search for plugins and "install" them on their instance? 

Think Stripe for example, 80% of the cases are probably the same when it comes to an integration with Stripe.
Or an email provider... Or a CRM... It's endless.

It would make pocketbase even more compelling to use than it already is!

Would love to know if that's on the cards at all?

**Comment by ganigeorgiev** [comment, marked as answer]
> but has there been a consideration to enable a marketplace of sorts that allows users of pocketbase to search for plugins and "install" them on their instance?

While PocketBase can be extended it requires a build process, we can't load code dynamically.
There are plans in the near future to allow binding to the Go app hooks via JS embedded interpreter ([goja](https://github.com/dop251/goja)), similar to the latest migration changes in v0.9, but initially the dynamic code will be executed only on application start. Changes while the application is running are a little more difficult due to various concurrency related edge cases.

Additionally, in order to allow fully qualified plugins support (eg. as in WordPress), we'll need also to be able to extend the Admin UI (see https://github.com/pocketbase/pocketbase/issues/898), which currently is not possible since it is a prebuilt static Svelte app.

The above may be added in the future, but for now this is out of the scope for v1. You can find a list with most of the currently planned tasks for v1 in https://github.com/orgs/pocketbase/projects/2.

**Reply by charbs-io** [reply]
Makes sense. Appreciate all the hard work, big fan of PB

**Reply by scriptcoded** [reply]
@ganigeorgiev have you seen how Hashicorp is dealing with this? https://github.com/hashicorp/go-plugin works around the need of building code on the fly 🙂

**Reply by ganigeorgiev** [reply]
@scriptcoded I've used the hashicorp plugin lib at work in the past, but it requires creating and maintaining a separate API and it will be hard to keep it in sync with the embedded Go one (the same apply for the wasm alternatives). As mentioned in the comment above, eventually we'll allow binding to the app hooks with JS via goja since the Go interop is almost seamless, but that will be handled after the filter enhancements and planned UI improvements (most likely in v0.13/v0.14).

**Reply by scriptcoded** [reply]
Sounds like you've thought about it thoroughly :) Maybe this is something you've thought about, but perhaps the interface towards goja could be kept rather generic? Perhaps so that something like go-plugin or another solution could be used in the future.

**Reply by ganigeorgiev** [reply]
@scriptcoded The wasm/rpc and the interpreter solutions are not compatible. The main benefit of using goja (or any other interpreter) is that it is part of your Go program and you can share Go structs, functions, etc. without having to create custom endpoints to serialize/deserialize the data so that you can move it from one context to another. Additionally, using an rpc plugin system doesn't just magically works, there is a lot of upfront configuration required (you'll need at least to setup the addresses and command patterns both sides has agreed to use) and it is very difficult to do that dynamically while the program is running.

**Reply by scriptcoded** [reply]
Okey, get it. Thanks for the explanation :)

**Reply by jimafisk** [reply]
We created an open source build tool to try and make adding Go plugins easier: https://github.com/pocketbuilds/xpb

Like Gani mentioned previously this still requires a build process, so you're not technically installing plugins on your current instance, you're rebuilding with the Plugin included and then replacing your binary with the new binary. The mechanism is different, but it should feel similar to installing plugins on Wordpress or modules on Drupal.

**Comment by rnbokade** [comment]
How about this, go has access to cli i.e. you can access cli programmatically from within a running go program. The plugin packages could be json/toml which defines steps to download the source code and dependencies and then configuring them to run as a seperate process. These packages can then either directly access the sqlite (there is a way for 2 processes to share sqlite afaik) or plugin talks to PB over admin apis. As for frontend, each plugin could expose their settings and metrics in file or maybe directly in db. The config page for particular plugin will fetch list of all keys and values to modify. Metrics page can do the same.

**Reply by ganigeorgiev** [reply]
Doing it this way doesn't eliminate the main problem which is that it requires maintaining 2 separate APIs and serialization format no matter what the communication between the separate processes is (eg. wasi/(g)rpc/http/...).

"Dynamically" loading server-side code will be supported with the next v0.17.0 release by using JavaScript event hooks binding via goja (_technically they will not be dynamically loaded since we trigger an app restart with `execve` but the startup is almost instantaneous_). The bindings will mirror the existing [Go event hooks](https://pocketbase.io/docs/event-hooks/) and will allow access to almost everything what the Go code can do with very little performance penalty (_for most of the time_). You can find a POC in the [`develop` branch](https://github.com/pocketbase/pocketbase/tree/develop). There will be also autogenerated ambient TypeScript declarations to assist with the codeintellisense (you can find some details on the documentation flow in https://github.com/dop251/goja/discussions/516).

Note that the above will allow only server-side "dynamic plugins" support. The Admin UI is still not extendable and need to be refactored to support it (this will be handled sometime in the future as part of https://github.com/pocketbase/pocketbase/issues/898).
